### PR TITLE
GH-38981: [R][Release] Don't update version.json on compatible version release

### LINKED ISSUE #38981: [R][Release] {docs/source/_static,r/pkgdown/assets}/versions.json have an invalid entry
### Describe the bug, including details regarding any error messages, version, and platform.

https://github.com/apache/arrow/commit/a9c99cfe362fc7d450711fed82f74e565f381337 added an entry for "14.0.0" to `docs/source/_static/versions.json` and `r/pkgdown/assets/versions.json` but they aren't needed because "14.0.0" and "14.0.1" are compatible version. So we don't have https://arrow.apache.org/docs/r/14.0/ yet.

CI is failed: https://github.com/apache/arrow/actions/runs/7019945363/job/19098561253

### Component(s)

R, Release

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

If we release a compatible version such as 14.0.0 -> 14.0.1 and 14.1.1 -> 14.1.2, we don't need to update `docs/source/_static/versions.json` and `r/pkgdown/assets/versions.json`. Because we use compatible version ("14.0" and "14.1") instead of exact version ("14.0.0" and "14.1.1").

### What changes are included in this PR?

Add more update condition to detect compatible version release.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* Closes: #38981

### CODE REVIEW

File: dev/release/utils-update-docs-versions.py [hunk 1]
```diff
@@ -76,14 +85,19 @@
...
+if release_type == "major" and split_version[1:] == ["0", "0"]:
+    # 14.0.0 -> 15.0.0
     new_r_versions = [
         {"name": f"{dev_r_version} (dev)", "version": "dev/"},
         {"name": f"{release_r_version} (release)", "version": ""},
         {"name": previous_r_name, "version": f"{previous_r_version}/"},
         *old_r_versions[2:],
     ]
 else:
+    # 14.0.1 -> 15.0.0
```
Comment (raulcd) [inline dev/release/utils-update-docs-versions.py:97]: isn't this case `release_type == "major" and split_version[1:] == ["0", "0"]` true due to release `15.0.0`? I am just confused about the comment.

**Review by raulcd — APPROVED** [review]
Thanks for this @kou !
Some minor comments only but happy with the overall change.

File: dev/release/utils-update-docs-versions.py [hunk 2]
```diff
@@ -38,7 +38,9 @@
 else:
     release_type = "patch"
 
-# Update main docs version script
+# Update main docs version script only when compatible version of the
+# stable version isn't change. Compatible version is ${MAJOR}.${MINOR}
```
[the remark this answers is not in the document]
Comment (kou) [reply dev/release/utils-update-docs-versions.py:42]: Oh, I wanted to write "isn't changed". Is it natural English?

```suggestion
# stable version isn't changed. Compatible version is ${MAJOR}.${MINOR}
```

File: dev/release/utils-update-docs-versions.py [hunk 3]
```diff
@@ -47,23 +49,30 @@
...
-    ]
-    with open(main_versions_path, 'w') as json_file:
-        json.dump(new_versions, json_file, indent=4)
-        json_file.write("\n")
+    # previous (compatible version) -> stable (compatible version)
+    #
+    # 13.Y.Z (13.Y) -> 14.0.0 (14.0): Update
+    # 14.0.0 (14.0) -> 14.0.1 (14.0): Not update
+    # 13.Y.Z (13.Y) -> 14.0.0 (14.0): Update
+    # 14.0.0 (14.0) -> 14.0.1 (14.0): Not update
```
[the remark this answers is not in the document]
Comment (kou) [reply dev/release/utils-update-docs-versions.py:57]: Oh, yes. You're right. But I wanted to use different order:

```suggestion
    # 13.Y.Z (13.Y) -> 14.0.0 (14.0): Update
    # 14.0.0 (14.0) -> 14.0.1 (14.0): Not update
    # 14.0.0 (14.0) -> 14.1.0 (14.1): Update
    # 14.0.1 (14.0) -> 14.1.0 (14.1): Update
```

File: dev/release/utils-update-docs-versions.py (hunk 1 above)
Comment (kou) [reply dev/release/utils-update-docs-versions.py:97]: It's false:

* `version = "14.0.1"`
* `split_version = ["14", "0", "1"]` <- `split_version[1:] == ["0", "1"]` not `["0", "0"]`
* `next_version = "15.0.0"`
* `split_next_version = ["15", "0", "0"]`
* `release_type = "major"`

File: dev/release/utils-update-docs-versions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kou) [reply dev/release/utils-update-docs-versions.py:42]: If "isn't changed" is natural, could you apply it and merge this PR?
If it's not natural, could you apply your suggestion and merge this PR?

[the remark this answers is not in the document]
Comment (raulcd) [reply dev/release/utils-update-docs-versions.py:42]: both sound good to me :) I am also non a native speaker :)

**Comment by raulcd** [comment]
I'll merge once CI passes. Thanks @kou for fixing this!

**Pull request merged**